File: Excel/300-399/315/315 Challenge.py
import pandas as pd
import re

path = "300-399/315/315 Letters Removal.xlsx"
input_df = pd.read_excel(path, usecols="A:B", nrows=9)
test = pd.read_excel(path, usecols="C", nrows=9)

planet_letters = "".join(
    sorted(set("".join(input_df["Planets"].dropna()).lower()))
)
pattern = re.compile(f"[{planet_letters}]", re.IGNORECASE)

result = (
    input_df["Author"]
    .apply(lambda s: re.sub(r"\s+", " ", pattern.sub("", s)).strip() or None)
    .rename("Answer Expected")
    .to_frame()
    .reset_index(drop=True)
)

print(result.equals(test))
# True

File: Excel/300-399/315/315 Challenge.R
library(tidyverse)
library(readxl)

path <- "300-399/315/315 Letters Removal.xlsx"
input <- read_excel(path, range = "A1:B10")
test <- read_excel(path, range = "C1:C10")

planet_letters <- input$Planets |>
  na.omit() |>
  tolower() |>
  paste(collapse = "") |>
  str_split("") |>
  unlist() |>
  unique() |>
  paste(collapse = "")

result <- input |>
  mutate(
    `Answer Expected` = Author |>
      str_remove_all(regex(
        paste0("[", planet_letters, "]"),
        ignore_case = TRUE
      )) |>
      str_squish() |>
      na_if("")
  ) |>
  select(`Answer Expected`)

all.equal(result, test)
# [1] TRUE
